src/v/archival/ntp_archiver_service.cc

```diff
@@ -1823,6 +1833,248 @@ ss::future<> ntp_archiver::housekeeping() {
+          "segments failed. Metadata for all remaining segments pending "
+          "deletion will be removed and these segments will have to be removed "
+          "manually.",
+          segments_to_remove.size(),
+          _max_segments_pending_deletion());
+    }
+    if (!all_deletes_succeeded && !backlog_size_exceeded) {
+        vlog(
+          _rtclog.info,
+          "Failed to delete all selected segments from cloud storage. Will "
+          "retry on the next housekeeping run.");
+    }
+    if (all_deletes_succeeded || backlog_size_exceeded) {
+        auto sync_timeout = config::shard_local_cfg()
+                              .cloud_storage_metadata_sync_timeout_ms.value();
+        auto deadline = ss::lowres_clock::now() + sync_timeout;
+        auto error = co_await _parent.archival_meta_stm()->cleanup_archive(
+          new_clean_offset, bytes_to_remove, deadline, _as);
+
+        if (error != cluster::errc::success) {
+            vlog(
+              _rtclog.info,
+              "Failed to clean up metadata after garbage collection: {}",
+              error);
+        } else {
+            // Remove manifests only if metadata no longer references them
+            for (const auto& path : manifests_to_remove) {
+                rem_batch.emplace_back(path);
+                if (rem_batch.size() >= batch_size) {
+                    std::vector<cloud_storage_clients::object_key> tmp;
+                    std::swap(tmp, rem_batch);
+                    co_await batch_delete(std::move(tmp));
+                }
+            }
+            co_await batch_delete(rem_batch);
+        }
+    }
+    _probe->segments_deleted(static_cast<int64_t>(successful_deletes));
+    vlog(
+      _rtclog.debug,
+      "Deleted {} spillover segments from the cloud",
+      successful_deletes);
+}
+
+ss::future<bool> ntp_archiver::batch_delete(
+  std::vector<cloud_storage_clients::object_key> keys) {
+    if (_remote.is_batch_delete_supported()) {
+        // Do batch delete, the batch size should be below the limit
+        auto res = co_await _remote.delete_objects(
+          get_bucket_name(), std::move(keys), _rtcnode);
+        if (res != cloud_storage::upload_result::success) {
+            vlog(_rtclog.error, "Failed to delete objects", res);
+            co_return false;
+        }
+    } else {
+        // Delete 1by1
+        size_t failures = 0;
+        co_await ss::max_concurrent_for_each(
+          keys,
+          _concurrency,
+          [this, &failures](cloud_storage_clients::object_key key) {
+              return _remote.delete_object(get_bucket_name(), key, _rtcnode)
+                .then([&failures](cloud_storage::upload_result res) {
+                    if (res != cloud_storage::upload_result::success) {
+                        failures++;
+                    }
+                });
+          });
+        if (failures > 0) {
+            co_return false;
+        }
+    }
```

Comment:
`remote::delete_objects` already provides an alternative implementation for backends with no plural delete. Is that not suitable?